src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,916 @@
+  const size_t nSmpFtrDim = (size_t)(nFtrDim * fraction);
+
+  for(size_t i = 0; i < numTree; ++i)
+  {
+    // Implement the logic for if data already exists.
+    // this is our new feature dimension.
+    MatType ftrs = arma::zeros(numPos + numNeg, nSmpFtrDim);
+
+    // effectively a 3d array.
+    MatType lbls = arma::zeros((numPos + numNeg ), gSize * gSize);
+    // still to be done: store features and labels calculated 
+    // in the loop and store it in these Matrices.
+    
+    size_t loop_iter = numImages;
+
+    // table is a vector which helps in converting Image from RGB2LUV.
+    double a, y0, maxi; 
+    a = std::pow(29.0, 3) / 27.0;
+    y0 = 8.0 / a;
+    maxi = 1.0 / 270.0;
+    arma::vec table(1064);
+    
+    for (size_t i = 0; i <= 1024; ++i)
+    {
+      table(i) = i / 1024.0;
+      
+      if (table(i) > y0)
+        table(i) = 116 * pow(table(i), 1.0/3.0) - 16.0;
+      else
+        table(i) = table(i) * a;
+
+      table(i) = table(i) * maxi;
+    }
+
+    for(size_t i = 1025; i < table.n_elem; ++i)
+      table(i) = table(i - 1);
+
+    size_t col_i = 0, col_s = 0, col_b = 0;
+    // process data of each image one by one.
+    for(size_t j = 0; j < loop_iter; ++j)
+    {
+      // these varaibles store image, boundaries and segmentation information
+      // for each image in our dataset.
+      MatType Img, bnds, segs;
+
+      Img = MatType(Images.colptr(col_i), colSize, rowSize * 3).t() / 255;
+      col_i += 3;
+      
+      bnds = MatType(Boundaries.colptr(col_b), colSize, rowSize).t();
+      col_b++;
+
+      segs = MatType(Segmentations.colptr(col_s), colSize, rowSize).t();
+      col_s++;
+
+      MatType mask(rowSize, colSize, arma::fill::ones);
+      mask.col(pRad - 1).fill(0);
+      mask.row( (mask.n_rows - 1) - (pRad - 1) ).fill(0);
+      mask.submat(0, 0, mask.n_rows - 1, pRad - 1).fill(0);
+      mask.submat(0, mask.n_cols - pRad, mask.n_rows - 1, 
+                  mask.n_cols - 1).fill(0);
+
+      // number of positive or negative patches per ground truth.
+      const size_t nPatchesPerGt = 500;
+
+      MatType dis;
+
+      // calculate distance transform of image boundary.
+      this->DistanceTransformImage(bnds, 1, dis);
+      // take square root for euclidean distance transform.
+      dis = arma::sqrt(dis);      
+
+      // find positive and negative edge locations using mask.
+      arma::uvec posLoc = arma::find( (dis < gRad) % mask );
+      arma::uvec negLoc = arma::find( (dis >= gRad) % mask );
+
+      // we take a random permutation of posLoc and negLoc.
+      posLoc = arma::shuffle(posLoc);
+      negLoc = arma::shuffle(negLoc);
+
+      size_t lenLoc = std::min((int) negLoc.n_elem, std::min((int) nPatchesPerGt,
+                              (int) posLoc.n_elem));
+      arma::umat loc(lenLoc * 2, 2);
+
+      for(size_t i = 0; i < lenLoc; ++i)
+        loc.row(i) = arma::ind2sub(arma::size(dis.n_rows, dis.n_cols), posLoc(i) ).t();
```

Comment:
Do you think, we could store the location in the columns, in this case we don't have to transpose the locations vector.